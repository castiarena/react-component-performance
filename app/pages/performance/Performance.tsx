import React, { Profiler, useState, useRef } from 'react';
import Layout from '../../components/Layout';
import ButtonAnchor from '../../components/ButtonAnchor';
import { Phar, SubTitle, Title } from '../../components/Typography';
import TableTest from '../test/components/TableTest';
import usePassengers from '../test/gateway/usePassengers';
import Box from '../../components/Box';
import FixedBox from '../../components/FixedBox';
import usePerformanceMetrics from './usePerformanceMetrics';

const isDevelopment = process.env.NODE_ENV === 'development';

const Performance = () => {
    const {
        passengers,
        handleLoadMorePassengers,
        fetchPassengers,
        passengerDuration,
    } = usePassengers();
    const [showPerformance, setShowPerformance] = useState(true);
    const {
        handlerRenderProfiler, name, phase, duration, time,
    } = usePerformanceMetrics();

    const handleClickFetchPassengers = async () => {
        await fetchPassengers();
    };

    const handleClickShowPerformance = () => {
        setShowPerformance(!showPerformance);
    };

    return (
        <Layout columns={1} gap={40}>
            <Title align="center" bold>
                Components performance measuring
            </Title>
            {showPerformance && isDevelopment && (
                <FixedBox>
                    <SubTitle bold ref={name} />
                    <Phar ref={phase} />
                    <Phar ref={duration} />
                    <Phar ref={time} />
                    <Phar>
                        Last request duration:
                        {passengerDuration}
                    </Phar>
                </FixedBox>
            )}
            <Layout columns={1} gap={10}>
                {!isDevelopment && (
                    <Layout columns={2} gap={10} fitted responsive>
                        <Box>
                            <SubTitle>Performance viewing</SubTitle>
                            <Phar>
                                In order to see the performance metrics running you should follow
                                the steps on the repository and run it locally, on development mode.
                            </Phar>
                            <ButtonAnchor
                                href="https://github.com/castiarena/react-component-performance"
                                target="_blank"
                            >
                                Repo
                            </ButtonAnchor>
                        </Box>
                        <Box shadow>
                            <SubTitle>Direct download</SubTitle>
                            <Phar>Download here the repository via https</Phar>
                            <ButtonAnchor
                                href="https://github.com/castiarena/react-component-performance.git"
                                download
                                kind="secondary"
                            >
                                Repo
                            </ButtonAnchor>
                        </Box>
                    </Layout>
                )}
                {isDevelopment && (
                    <>
                        <Layout columns={2} gap={10} responsive>
                            <Box fitted>
                                <Phar>
                                    With the next button you can fetch the data from the API
                                </Phar>
                                <ButtonAnchor onClick={handleClickFetchPassengers}>
                                    Fetch it!
                                </ButtonAnchor>
                            </Box>

                            <Box fitted>
                                <Phar>
                                    And whit this button you can see the performance metrics of the
                                    first, render and the repainting on each update, by using
                                    React.Profiler component.
                                </Phar>
                                <ButtonAnchor onClick={handleClickShowPerformance} kind="secondary">
                                    Toggle performance tab
                                </ButtonAnchor>
                            </Box>
                        </Layout>
                        <Profiler id="Passengers" onRender={handlerRenderProfiler}>
                            <TableTest
                                items={passengers}
                                handleLoadMoreItems={handleLoadMorePassengers}
                            />
                        </Profiler>
                    </>
                )}
            </Layout>
        </Layout>
    );
};

export default Performance;
