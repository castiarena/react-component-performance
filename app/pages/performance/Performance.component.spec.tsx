import React from 'react';
import 'jest-styled-components';
import { BrowserRouter } from 'react-router-dom';
import Performance from './Performance';
import renderWithTheme from '../../testing-tools/renderWithTheme';

describe('Performance component test suite', () => {
    it('smoke test, snapshot only ', () => {
        const { getByText } = renderWithTheme(
            <BrowserRouter>
                <Performance />
            </BrowserRouter>,
        );
        expect(getByText('Components performance measuring')).toBeDefined();
    });
});
