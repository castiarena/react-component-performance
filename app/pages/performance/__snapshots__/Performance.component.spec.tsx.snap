// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance component test suite smoke test, snapshot only  1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-template-columns: repeat(1,1fr);
  grid-column-gap: 40px;
  padding: 20px;
}

.c1 {
  border: none;
  padding: 16px 30px;
  border-radius: 10px;
  background-color: #0084ff;
  color: #fff;
  -webkit-transition: background-color .2s ease-out;
  transition: background-color .2s ease-out;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:hover {
  cursor: pointer;
  background-color: #a7c3dc;
}

.c1:active,
.c1:focus {
  outline: none;
}

<div
    class="c0"
  >
    <div>
      <h1>
        Performance
      </h1>
      <p>
        <a
          class="c1"
          download=""
          href="./assets/instructions.pdf"
        >
          Download instructions
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
